dbms/src/Storages/DeltaMerge/RestoreDMFile.cpp

```diff
@@ -29,38 +31,61 @@ DMFilePtr restoreDMFileFromRemoteDataSource(
     UInt64 file_page_id,
     UInt64 meta_version)
 {
-    auto path_delegate = dm_context.path_pool->getStableDiskDelegator();
-    auto wn_ps = dm_context.global_context.getWriteNodePageStorage();
-    auto full_page_id = UniversalPageIdFormat::toFullPageId(
-        UniversalPageIdFormat::toFullPrefix(dm_context.keyspace_id, StorageType::Data, dm_context.physical_table_id),
-        file_page_id);
-    auto full_external_id = wn_ps->getNormalPageId(full_page_id);
-    auto local_external_id = UniversalPageIdFormat::getU64ID(full_external_id);
-    auto remote_data_location = wn_ps->getCheckpointLocation(full_page_id);
-    const auto & lock_key_view = S3::S3FilenameView::fromKey(*(remote_data_location->data_file_id));
-    auto file_oid = lock_key_view.asDataFile().getDMFileOID();
-    auto prepared = remote_data_store->prepareDMFile(file_oid, file_page_id);
-    auto dmfile = prepared->restore(DMFileMeta::ReadMode::all(), meta_version);
-    // gc only begin to run after restore so we can safely call addRemoteDTFileIfNotExists here
-    path_delegate.addRemoteDTFileIfNotExists(local_external_id, dmfile->getBytesOnDisk());
+    DMFilePtr dmfile;
+    try
+    {
+        auto path_delegate = dm_context.path_pool->getStableDiskDelegator();
+        auto wn_ps = dm_context.global_context.getWriteNodePageStorage();
+        auto full_page_id = UniversalPageIdFormat::toFullPageId(
+            UniversalPageIdFormat::toFullPrefix(
+                dm_context.keyspace_id,
+                StorageType::Data,
+                dm_context.physical_table_id),
+            file_page_id);
+        auto full_external_id = wn_ps->getNormalPageId(full_page_id);
+        auto local_external_id = UniversalPageIdFormat::getU64ID(full_external_id);
+        auto remote_data_location = wn_ps->getCheckpointLocation(full_page_id);
+        const auto & lock_key_view = S3::S3FilenameView::fromKey(*(remote_data_location->data_file_id));
+        auto file_oid = lock_key_view.asDataFile().getDMFileOID();
+        auto prepared = remote_data_store->prepareDMFile(file_oid, file_page_id);
+        dmfile = prepared->restore(DMFileMeta::ReadMode::all(), meta_version);
+        // gc only begin to run after restore so we can safely call addRemoteDTFileIfNotExists here
+        path_delegate.addRemoteDTFileIfNotExists(local_external_id, dmfile->getBytesOnDisk());
+    }
+    catch (DB::Exception & e)
+    {
+        e.addMessage(fmt::format("file_page_id={} meta_version={}", file_page_id, meta_version));
+        e.rethrow();
+    }
```

Comment:
Where does the Exception come from?

Reply:
when trying to use an old version tiflash binary but load new data type
```
[2024/12/26 23:27:43.778 +08:00] [ERROR] [Exception.cpp:91] ["Storage inited fail, keyspace=4294967295 table_id=127: Code: 50, e.displayText() = DB::Exception: Unknown data type family: StringV2: file_page_id=14494 meta_version=0: while restoreSegment, segment_id=7997 ident=keyspace=4294967295 table_id=127: 
  0x55a0533503d3    StackTrace::StackTrace() [tiflash+35009491]
                    dbms/src/Common/StackTrace.cpp:23
  0x55a05334daa6    DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int) [tiflash+34998950]
                    dbms/src/Common/Exception.h:46
  0x55a058b076db    DB::DataTypeFactory::get(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::IAST> const&) const [tiflash+126985947]
                    dbms/src/DataTypes/DataTypeFactory.cpp:119
  0x55a058b0718b    DB::DataTypeFactory::get(std::__1::shared_ptr<DB::IAST> const&) const [tiflash+126984587]
                    dbms/src/DataTypes/DataTypeFactory.cpp:87
  0x55a058b0740a    DB::DataTypeFactory::getOrSet(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) [tiflash+126985226]
                    dbms/src/DataTypes/DataTypeFactory.cpp:56
  0x55a0588d4228    DB::DM::ColumnStat::mergeFromProto(dtpb::ColumnStat const&) [tiflash+124678696]
                    dbms/src/Storages/DeltaMerge/File/ColumnStat.h:81
  0x55a0588d0f90    DB::DM::DMFileMetaV2::parseExtendColumnStat(std::__1::basic_string_view<char, std::__1::char_traits<char>>) [tiflash+124665744]
                    dbms/src/Storages/DeltaMerge/File/DMFileMetaV2.cpp:127
  0x55a0588d04b3    DB::DM::DMFileMetaV2::parse(std::__1::basic_string_view<char, std::__1::char_traits<char>>) [tiflash+124662963]
                    dbms/src/Storages/DeltaMerge/File/DMFileMetaV2.cpp:86
  0x55a0588d2771    DB::DM::DMFileMetaV2::read(std::__1::shared_ptr<DB::FileProvider> const&, DB::DM::DMFileMeta::ReadMode const&) [tiflash+124671857]
                    dbms/src/Storages/DeltaMerge/File/DMFileMetaV2.cpp:292
  0x55a0588af57e    DB::DM::DMFile::restore(std::__1::shared_ptr<DB::FileProvider> const&, unsigned long, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::DM::DMFileMeta::ReadMode const&, unsigned long, unsigned int) [tiflash+124527998]
                    dbms/src/Storages/DeltaMerge/File/DMFile.cpp:0
  0x55a058862b25    DB::DM::restoreDMFileFromLocal(DB::DM::DMContext const&, unsigned long, unsigned long) [tiflash+124214053]
                    dbms/src/Storages/DeltaMerge/RestoreDMFile.cpp:72
  0x55a058859601    DB::DM::StableValueSpace::restore(DB::DM::DMContext&, DB::ReadBuffer&, unsigned long) [tiflash+124175873]
                    dbms/src/Storages/DeltaMerge/StableValueSpace.cpp:197
  0x55a058859456    DB::DM::StableValueSpace::restore(DB::DM::DMContext&, unsigned long) [tiflash+124175446]
                    dbms/src/Storages/DeltaMerge/StableValueSpace.cpp:182
  0x55a05879d494    DB::DM::Segment::restoreSegment(std::__1::shared_ptr<DB::Logger> const&, DB::DM::DMContext&, unsigned long) [tiflash+123405460]
                    dbms/src/Storages/DeltaMerge/Segment.cpp:429
  0x55a058732c07    std::__1::__function::__func<DB::DM::DeltaMergeStore::DeltaMergeStore(DB::Context&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned int, long, long, bool, std::__1::vector<DB::DM::ColumnDefine, std::__1::allocator<DB::DM::ColumnDefine>> const&, DB::DM::ColumnDefine const&, bool, unsigned long, std::__1::shared_ptr<std::__1::vector<DB::DM::LocalIndexInfo, std::__1::allocator<DB::DM::LocalIndexInfo>>>, DB::DM::DeltaMergeStore::Settings const&, DB::ThreadPoolImpl<DB::ThreadFromGlobalPoolImpl<false>>*)::$_1, std::__1::allocator<DB::DM::DeltaMergeStore::DeltaMergeStore(DB::Context&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned int, long, long, bool, std::__1::vector<DB::DM::ColumnDefine, std::__1::allocator<DB::DM::ColumnDefine>> const&, DB::DM::ColumnDefine const&, bool, unsigned long, std::__1::shared_ptr<std::__1::vector<DB::DM::LocalIndexInfo, std::__1::allocator<DB::DM::LocalIndexInfo>>>, DB::DM::DeltaMergeStore::Settings const&, DB::ThreadPoolImpl<DB::ThreadFromGlobalPoolImpl<false>>*)::$_1>, void ()>::operator()() [tiflash+122969095]
                    dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:295
  0x55a0534088f5    std::__1::packaged_task<void ()>::operator()() [tiflash+35764469]
                    /DATA/disk1/ra_common/tiflash-env-17/sysroot/bin/../include/c++/v1/future:1891
  0x55a053406e05    DB::ThreadPoolImpl<DB::ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<DB::ThreadFromGlobalPoolImpl<false>, void*>) [tiflash+35757573]
                    /DATA/disk1/ra_common/tiflash-env-17/sysroot/bin/../include/c++/v1/__functional/function.h:517
  0x55a053409723    std::__1::__function::__func<DB::ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void DB::ThreadPoolImpl<DB::ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), std::__1::allocator<DB::ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void DB::ThreadPoolImpl<DB::ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()>, void ()>::operator()() [tiflash+35768099]
                    dbms/src/Common/UniThreadPool.cpp:167
  0x55a0534055f5    DB::ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) [tiflash+35751413]
                    /DATA/disk1/ra_common/tiflash-env-17/sysroot/bin/../include/c++/v1/__functional/function.h:517
  0x55a053407cf5    void* std::__1::__thread_proxy[abi:ue170006]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void DB::ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) [tiflash+35761397]
                    dbms/src/Common/UniThreadPool.cpp:167
  0x7fc92f140802    start_thread [libc.so.6+653314]
  0x7fc92f0e0450    clone3 [libc.so.6+259152]
```